jb/itch5/array_based_order_book.hpp

```diff
@@ -0,0 +1,660 @@
+    if (top_levels_.at(rel_px) == 0) {
+      // gets the new inside (if any)
+      px_inside_ = next_best_price();
+      if (px_inside_ == empty_quote().first) {
+        // last top_levels_ price was removed...
+        // ... get the new inside from the bottom_levels
+        if (not bottom_levels_.empty()) {
+          auto new_inside_it = bottom_levels_.begin();
+          px_inside_ = new_inside_it->first;
+        }
+        // redefine limits
+        auto limits = get_limit_top_prices(px_inside_);
+        px_begin_top_ = std::get<0>(limits);
+        px_end_top_ = std::get<1>(limits);
+        // move tail prices from bottom_levels (px_begin_top_ {, .begin()})
+        move_bottom_to_top();
+      }
+    }
+    return true; // it is an inside change
+  }
+
+  /**
+   * Testing hook.
+   * @returns true is px1 < px2
+   * To test different implementations for buy and sell sides
+   */
+  bool price_comp(price4_t const& px1, price4_t const& px2) const {
+    return better_(px1, px2);
+  }
+
+  /** Testing hook.
+   * @returns {px_begin_top_, px_end_top_}
+   */
+  auto get_limits() const {
+    return std::make_tuple(px_begin_top_, px_end_top_);
+  }
+
+  /** Testing hook.
+   * @returns {bottom_levels_ size , top_levels_ size}
+   */
+  auto get_counts() const {
+    return std::make_pair(bottom_levels_.size(), top_levels_count());
+  }
+
+  /** Testing hooks to increase coverage
+   */
+  auto test_price_to_relative(price4_t const& px) const {
+    return price_to_relative(px);
+  }
+  // to test when is an empty side
+  auto test_relative_worst_top_level() const {
+    return relative_worst_top_level();
+  }
+  // to test with a px worse than px_begin_top
+  void test_move_top_to_bottom(price4_t const& px) {
+    move_top_to_bottom(px);
+  }
+
+private:
+  /// The value used to represent an empty bid
+  static half_quote empty_bid() {
+    return half_quote(price4_t(0), 0);
+  }
+
+  /// The value used to represent an empty offer
+  static half_quote empty_offer() {
+    return half_quote(max_price_field_value<price4_t>(), 0);
+  }
+
+  /**
+   * Transforms a top_levels_ relative position into an absolute price.
+   * @param rel relative position of a price at top_levels_
+   * @returns the absolute price
+   */
+  price4_t relative_to_price(int rel) const {
+    int rel_tick = (better_(price4_t(1), price4_t(0))) ? rel : max_size_ - rel;
+    price4_t px_base =
+        (better_(price4_t(1), price4_t(0))) ? px_begin_top_ : px_end_top_;
+    return relative_above_base(px_base, rel_tick);
+  }
+
+  /**
+   * @param px_base base absolute price
+   * @param rel_tick price levels above px_base
+   * @returns absolute price rel_tick price levels above px_base
+   */
+  price4_t
+  relative_above_base(price4_t const& px_base, int const rel_tick) const {
+    int const wire_max = max_price_field_value<price4_t>().as_integer();
+
+    // check if px_base is above $1.00
+    if (px_base > price4_t(TK_DOLLAR)) {
```

Comment:
This logic about when we switch from penny increments to sub-penny increments is what price_levels() is supposed to handle.  Why do you need to reimplement it?